### PR TITLE
Fix: WebSockets JS API refactoring

### LINKED ISSUE #516: `WebSocketApiImplementation` allocations
Input byte array is not pooled because `SendAsync` API should be modified to support `Memory<byte>` instead of `byte[]` (otherwise we can't use `ArrayPool`)

```
var byteArray = new byte[binaryData.Count];
                    for (var i = 0; i < binaryData.Count; i++) byteArray[i] = Convert.ToByte(binaryData[i]);

                    payloadListScope.Value.Add(byteArray);

                    var payload = new Payload
                    {
                        Bytes = payloadListScope.Value,
                    };

                    await webSocket.SendAsync(payload, ct);
```

ByteArray for Receive is not pooled either, it should be done similarly to `EngineAPI`

```
void ReceivedHandler(IMessage message)
            {
                try
                {
                    object result;

                    if (message is BinaryMessage binaryMessage)
                    {
                        byte[] bytes = binaryMessage.IncomingBytes.SelectMany(x => x).ToArray();
                        result = new { type = "Binary", data = bytes };
                    }
                    else if (message is DefaultTextMessage defaultMessage) { result = new { type = "Text", data = defaultMessage.Message }; }
                    else { throw new NotSupportedException($"Unsupported message type: {message.GetType().Name}"); }

                    tcs.TrySetResult(result);
                }
```

### PR DESCRIPTION
**Pull request by mikhail-dcl**
### Impossibility to export types properly

It was not possible to export the type like this:
```
module.exports.WebSocket = WebSocket
```

as JS complained about `constructor`:
> Class constructor WebSocket cannot be invoked without 'new' 

We had to do it via a function: 
```
module.exports.WebSocket = (url, protocols) => new WebSocket(url, protocols)
```

It's improper because on the consumer side we lose the original type as it refers exclusively to lambda.

**Cause**

`Validates` API wraps every type/function in its own type, thus on the user-end the knowledge about the original type is lost.

**Fix**

Commented everything related to `Validates`. In the future we need to have the possibility to activate it for Debug/Test purposes only

### Non-functional WebSocket API

**Symptoms**

All scenes using WebSocket API do not work.

#1236 
#935

**Cause**

- The implementation used `SocketIO` that no scenes have knowledge about, thus there was a protocol mismatch.
- Several mistakes in `WebSocketApi.js`
- Wrong marshaling of data between JS and ClearScript in both directions
- Impossibility to export `WebSocket` properly led to absence of callbacks set

**Fix**

- Fresh clean implementation based on `ClientWebSocket` solely
- Various fixes to `WebSocketApi.js`
- Fixes #516 as all allocations are under our direct control and were properly handled

## Remaining issues

`ColyseusSDK` breaks connection sporadically on the host side at `145,60` (Metamall).

```
SocketException: An established connection was aborted by the software in your host machine.
```

**Possible reason**

- Scene processes received data for too long. It is clearly visible due to logs and how delayed the instantiation of coins/red rocks is.
- `WebSocket` communication times out on the server side
- `WebSocket` is aborted without a handshake
- The scene is still non-functional 😢 

## How to test

- Genesis Plaza dialogs should function properly 🕺 , there should be no errors related to the WebSockets
- There is a chance that Metamall (`145,60`) can function as well under certain circumstances...

### CODE REVIEW

File: Explorer/Assets/Scripts/CrdtEcsBridge/JsModulesImplementation/ClientWebSocketApiImplementation.cs [hunk 1]
```diff
@@ -0,0 +1,225 @@
...
+        }
+
+        public async UniTask SendBinaryAsync(int websocketId, IArrayBuffer data, CancellationToken ct)
+        {
+            WebSocketRental webSocket = GetInstanceOrThrow(websocketId);
+
+            var bytesCount = (int)data.Size;
+            if (bytesCount == 0) return;
+
+            using PoolableByteArray poolableArray = instancePoolsProvider.GetCrdtRawDataPool(bytesCount);
```
Comment (m3taphysics) [inline Explorer/Assets/Scripts/CrdtEcsBridge/JsModulesImplementation/ClientWebSocketApiImplementation.cs:63]: Is this the CrdtRawDataPool being used in a non-crdt context? So it is more generic? Maybe we can just rename it for clarity?

File: Explorer/Assets/Scripts/CrdtEcsBridge/JsModulesImplementation/ClientWebSocketApiImplementation.cs [hunk 2]
```diff
@@ -0,0 +1,225 @@
...
+                throw new ArgumentException($"WebSocket with id {websocketId} does not exist.");
+
+            await webSocket.WebSocket.CloseAsync(WebSocketCloseStatus.NormalClosure, string.Empty, ct);
+        }
+
+        public async UniTask<IWebSocketApi.ReceiveResponse> ReceiveAsync(int websocketId, CancellationToken ct)
+        {
+            WebSocketRental webSocket = GetInstanceOrThrow(websocketId);
+
+            (PoolableByteArray result, WebSocketMessageType messageType, WebSocketCloseStatus closeStatus)
```
Comment (m3taphysics) [inline Explorer/Assets/Scripts/CrdtEcsBridge/JsModulesImplementation/ClientWebSocketApiImplementation.cs:95]: Can receive throw and then we have a floating instance of WebSocket from the pool?

File: Explorer/Assets/Scripts/CrdtEcsBridge/JsModulesImplementation/ClientWebSocketApiImplementation.cs [hunk 3]
```diff
@@ -0,0 +1,225 @@
...
+                        bool endOfMessage = pages - 1 == i;
+                        await rental.WebSocket.SendAsync(subBuffer, messageType, endOfMessage, ct).ConfigureAwait(false);
+                    }
+                }
+                finally { rental.SendLock.Release(); }
+            }
+
+            public async Task<(PoolableByteArray result, WebSocketMessageType messageType, WebSocketCloseStatus closeStatus)> ReceiveAsync(ClientWebSocket webSocket, IInstancePoolsProvider instancePoolsProvider, CancellationToken ct)
+            {
+                PoolableByteArray finalBuffer = PoolableByteArray.EMPTY;
```
Comment (m3taphysics) [inline Explorer/Assets/Scripts/CrdtEcsBridge/JsModulesImplementation/ClientWebSocketApiImplementation.cs:184]: Do we need to ensure the buffer is disposed if there is any exception?

**Review by m3taphysics — APPROVED** [review]

File: Explorer/Assets/Scripts/CrdtEcsBridge/JsModulesImplementation/ClientWebSocketApiImplementation.cs (hunk 2 above)
Comment (mikhail-dcl) [reply Explorer/Assets/Scripts/CrdtEcsBridge/JsModulesImplementation/ClientWebSocketApiImplementation.cs:95]: We don't pool WebSockets - according to the .NET API they are not reusable

**Review by sandrade-dcl — APPROVED** [review]

**Comment by LucasLioyQA** [comment]
**MAC testing**: 

IA dialogues:  they are all triggered now:

simone dialogue
Aisha dialogue
DOgue dialogue

However, buttons don't respond and when writing questions to interact with these NPCs, if using keys which perform actions at UI level, those actions are performed, when they shouldn't. Check video 

Screenshot 2024-06-27 at 3 36 07 PM

Then the build froze and broke the bar scene only twice: once when interacting with Simone and another time when interacting with Aisha. Here are the logs: 

[bar scene breaks after interacting with woman NPC with IA in the bar.log](https://github.com/user-attachments/files/16019721/bar.scene.breaks.after.interacting.with.woman.NPC.with.IA.in.the.bar.log)
[freeze while interacting with simone asking questions.log](https://github.com/user-attachments/files/16019722/freeze.while.interacting.with.simone.asking.questions.log)


Then, dialogues with other NPCs look ok, with the exception of long  dialogues where the press F to skip overlaps with text sometimes: 

Screenshot 2024-06-27 at 3 36 39 PM
Screenshot 2024-06-27 at 3 35 41 PM

This issue will be reported separately , too.

Finally, MetaMall. Usually it works ok and coins can be obtained and coin counter works ok. Check video

However, sometimes the scene seems to break and coins cannot be obtained . Check video

**Pull request merged**